cql3/expr/expression.cc

```diff
@@ -1310,29 +1310,7 @@ constant evaluate(term* term_ptr, const query_options& options) {
         return constant::make_null();
     }
 
-    if (dynamic_cast<abstract_marker*>(term_ptr) != nullptr
-    || dynamic_cast<tuples::value*>(term_ptr) != nullptr
-    || dynamic_cast<tuples::delayed_value*>(term_ptr) != nullptr
-    || dynamic_cast<lists::value*>(term_ptr) != nullptr
-    || dynamic_cast<lists::delayed_value*>(term_ptr) != nullptr 
-    || dynamic_cast<sets::value*>(term_ptr) != nullptr
-    || dynamic_cast<sets::delayed_value*>(term_ptr) != nullptr
-    || dynamic_cast<maps::value*>(term_ptr) != nullptr
-    || dynamic_cast<maps::delayed_value*>(term_ptr) != nullptr
-    || dynamic_cast<user_types::value*>(term_ptr) != nullptr
-    || dynamic_cast<user_types::delayed_value*>(term_ptr) != nullptr
-    || dynamic_cast<functions::function_call*>(term_ptr) != nullptr) {
-        return evaluate(term_ptr->to_expression(), options);
-    }
```

Comment:
I don't see removals of the old code. Is it still reachable?

In general the best way to construct such series is to pair new code + removal of old code in the same patch. This way the reviewer can match old/new code, and can see we don't increase bloat.

Reply:
Yes old `bind()` implementations are still reachable, because [`prepare_term` uses `bind()` to return a `terminal`](https://github.com/scylladb/scylla/blob/dd9d6c081ee48e670eda70e14a765442db62c0ec/cql3/expr/term_expr.cc#L410)
It can't be replaced with `evaluate` because `expr::constant` is not convertible to `terminal`.

Old code will be removed along with `term` in the next PR.

Reply:
I see. Maybe it's better to merge the two series. It's better to have two large series than one very large series, but in this case we have a performance regression (or maybe not, we can measure) plus it leaves a lot of duplicate code.

I think we could have preserved performance by having expr::term_expr and expr_term : term, and converting between them during prepare. But let's not rewrite the whole series.